api/opencl_architecture.asciidoc

```diff
@@ -198,10 +198,12 @@ A command submitted to a device will not launch until prerequisites that
 constrain the order of commands have been resolved.
 These prerequisites have three sources:
 
-  * They may arise from commands submitted to a command-queue that constrain
-    the order in which commands are launched.
-    For example, commands that follow a command-queue barrier will not
-    launch until all commands prior to the barrier are complete.
+  * The first source of prerequisites is implicit dependencies between commands
+    enqueued to the same command-queue which arise as follows:
+  ** Commands enqueued after a command-queue barrier have this barrier command
+     a prerequisite.
```

Comment:
To clarify, i'm reading this as *all* barrier commands in a command-queue are a prerequisite of the new command.
i.e. if you had the following
A. clEnqueueBarrier()
B. clEnqueueBarrier()
C. clEnqueueNDRangeKernel()

Then the dependency graph would be:
```
A 
| \
B  |
| /
C
```

Rather than if a prerequisite is just the last barrier added to the queue

```
A
|
B
|
C
```

There's no difference in the command scheduling, so this is nitpicking a bit.

Reply:
I agree, though I don't think the difference is observable.  In the first case C has an explicit dependency on A and B, and in the second case there is an explicit dependency on B and an implicit (transitive) dependency on B.  Since there's no way to query the dependencies for a command, I don't think it really matters which interpretation gets used.

FWIW, I think the mental model is simpler when it just considers the last barrier before the command (so, the second interpretation).  If we want to describe it to remove ambiguity and to nudge towards this mental model I think we could adopt the "slightly more verbose" version I suggested in a comment above, but I don't think this is strictly necessary.

Reply:
It shouldn't matter but better not to use something ambiguous. I've changed the wording to say "the preceding barrier command" instead of "this barrier command" and added the missing "as".